spatial/r3/transform.go

```diff
@@ -0,0 +1,267 @@
+
+package r3
+
+import (
+	"math"
+)
+
+// Transform represents a 3D spatial transformation.
+// The zero value of Transform is the identity transform.
+type Transform struct {
+	// in order to make the zero value of Transform represent the identity
+	// transform we store it with the identity matrix subtracted.
+	// These diagonal elements are subtracted such that
+	//  d00 = x00-1, d11 = x11-1, d22 = x22-1, d33 = x33-1
+	// where x00, x11, x22, x33 are the matrix diagonal elements.
+	// We can then check for identity in if blocks like so:
+	//  if T == (Transform{})
+	d00, x01, x02, x03 float64
+	x10, d11, x12, x13 float64
+	x20, x21, d22, x23 float64
+	x30, x31, x32, d33 float64
+}
+
+// Transform applies the Transform to the argument vector
+// and returns the result.
+func (t Transform) Transform(v Vec) Vec {
+	// https://github.com/mrdoob/three.js/blob/dev/src/math/Vector3.js#L262
+	w := 1 / (t.x30*v.X + t.x31*v.Y + t.x32*v.Z + t.d33 + 1)
+	return Vec{
+		X: ((t.d00+1)*v.X + t.x01*v.Y + t.x02*v.Z + t.x03) * w,
+		Y: (t.x10*v.X + (t.d11+1)*v.Y + t.x12*v.Z + t.x13) * w,
+		Z: (t.x20*v.X + t.x21*v.Y + (t.d22+1)*v.Z + t.x23) * w,
+	}
+}
+
+// zeroTransform is the Transform that returns zeroTransform when multiplied by any Transform.
+var zeroTransform = Transform{d00: -1, d11: -1, d22: -1, d33: -1}
+
+// NewTransform returns a new Transform type and populates its elements
+// with values passed in row-major form. If val is nil then NewTransform
+// returns a Transform filled with zeros.
+func NewTransform(a []float64) Transform {
+	if a == nil {
+		return zeroTransform
+	}
+	if len(a) != 16 {
+		panic("Transform is initialized with 16 values")
+	}
+	return Transform{
+		d00: a[0] - 1, x01: a[1], x02: a[2], x03: a[3],
+		x10: a[4], d11: a[5] - 1, x12: a[6], x13: a[7],
+		x20: a[8], x21: a[9], d22: a[10] - 1, x23: a[11],
+		x30: a[12], x31: a[13], x32: a[14], d33: a[15] - 1,
+	}
+}
+
+// ComposeTransform creates a new transform for a given translation to
+// positon, scaling vector scale and quaternion rotation.
+// The identity Transform is constructed with
+//  ComposeTransform(Vec{}, Vec{1,1,1}, Rotation{})
+func ComposeTransform(position, scale Vec, q Rotation) Transform {
+	x2 := q.Imag + q.Imag
+	y2 := q.Jmag + q.Jmag
+	z2 := q.Kmag + q.Kmag
+	xx := q.Imag * x2
+	yy := q.Jmag * y2
+	zz := q.Kmag * z2
+	xy := q.Imag * y2
+	xz := q.Imag * z2
+	yz := q.Jmag * z2
+	wx := q.Real * x2
+	wy := q.Real * y2
+	wz := q.Real * z2
+
+	var t Transform
+	t.d00 = (1-(yy+zz))*scale.X - 1
+	t.x10 = (xy + wz) * scale.X
+	t.x20 = (xz - wy) * scale.X
+
+	t.x01 = (xy - wz) * scale.Y
+	t.d11 = (1-(xx+zz))*scale.Y - 1
+	t.x21 = (yz + wx) * scale.Y
+
+	t.x02 = (xz + wy) * scale.Z
+	t.x12 = (yz - wx) * scale.Z
+	t.d22 = (1-(xx+yy))*scale.Z - 1
+
+	t.x03 = position.X
+	t.x13 = position.Y
+	t.x23 = position.Z
+	return t
+}
+
+// Translate adds Vec to the positional Transform.
+func (t Transform) Translate(v Vec) Transform {
```

Comment:
`Add`?

Reply:
Gioui uses [Offset](https://pkg.go.dev/gioui.org/f32#Affine2D.Offset) and [three.js uses Translation](https://threejs.org/docs/index.html?q=matrix#api/en/math/Matrix4.makeTranslation) and [setPosition](https://threejs.org/docs/index.html?q=matrix#api/en/math/Matrix4.setPosition). I think there is a preference to not use `Add` in the industry for this operation since it lends itself to be confused with ordinary matrix additon.

Reply:
And yet the docs use the "add" to describe the action it takes. I would call it add and explain in the docs with reference to translate and/or offset. The type system lets the user know that it's not ordinary matrix addiition, though note that really, it is, just with a different in-memory representation of a translation-only transform matrix.

Reply:
I guess I can convince myself `Add` is fine. If I were a new user to gonum I think I'd be confused by the `Add` method and wonder how I change the Translate part of the transform but hey, it should be easy enough to get used to. Will change it

Reply:
Another question, what method do we use to set the positional part of the transform to a known value?

Reply:
`SetPosition`?